src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -64,15 +64,31 @@ internal sealed class HttpConnectionPool : IDisposable
         /// <summary>The time, in milliseconds, that an authority should remain in <see cref="_altSvcBlocklist"/>.</summary>
         private const int AltSvcBlocklistTimeoutInMilliseconds = 10 * 60 * 1000;
 
-        /// <summary>List of idle connections stored in the pool.</summary>
-        private readonly List<CachedConnection> _idleConnections = new List<CachedConnection>();
-        /// <summary>The maximum number of connections allowed to be associated with the pool.</summary>
-        private readonly int _maxConnections;
+        // HTTP/1.1 connection pool
+
+        /// <summary>List of available HTTP/1.1 connections stored in the pool.</summary>
+        private readonly List<HttpConnection> _availableHttp11Connections = new List<HttpConnection>();
+        /// <summary>The maximum number of HTTP/1.1 connections allowed to be associated with the pool.</summary>
+        private readonly int _maxHttp11Connections;
+        /// <summary>The number of HTTP/1.1 connections associated with the pool, including in use, available, and pending.</summary>
+        private int _associatedHttp11ConnectionCount;
+        /// <summary>The number of HTTP/1.1 connections that are in the process of being established.</summary>
+        private int _pendingHttp11ConnectionCount;
+        /// <summary>Queue of requests waiting for an HTTP/1.1 connection.</summary>
+        private RequestQueue<HttpConnection> _http11RequestQueue;
+
+        // HTTP/2 connection pool
+
+        /// <summary>List of available HTTP/2 connections stored in the pool.</summary>
+        private List<Http2Connection>? _availableHttp2Connections;
```

Comment:
Seeing duplicate fields and methods for HTTP 1.1 and HTTP 2 (`_http?RequestQueue`, `_associatedHttp?ConnectionCount`, `_availableHttp?Connections`, `AddHttp?ConnectionAsync`, `CheckForHttp?ConnectionInjection`, `GetHttp?ConnectionAsync`, ...) Wouldn't it be worth it to extract it into a generic class and thus reduce the amount of code in here?

This is just a clean-up suggestion, it's not a prereq to get this PR merged. Perhaps you already have some ideas in your mind...

Reply:
I thought about trying to unify these, but in the end I didn't. Or rather, I decided to share `RequestQueue<T>` here but not the rest of the fields. 

There are several important differences here between HTTP/1.1 and HTTP2/3:

(1) HTTP/1.1 has a (optional) hard connection limit, which we don't have for HTTP2 or HTTP3 (but perhaps we should).
(2) For HTTP/1.1, we allow multiple pending connections at a time, whereas for HTTP2/3 we only allow one pending connection at a time.
(3) We know that HTTP/1.1 connections can only handle a single request at a time, so we always remove them from the available list when we are about to use them. For HTTP2/3, we only remove them from the available list if they can't handle any more requests at the moment.

Etc...

All that said, I think there is more potential unification that could happen here; I will look at this and see what makes sense to do.